key_action_map.cpp

```diff
@@ -2,157 +2,195 @@
-        assert(m.to_action(sf::Keyboard::D) == action_type::turn_right);
-        assert(m.to_action(sf::Keyboard::W) == action_type::accelerate);
-        assert(m.to_action(sf::Keyboard::S) == action_type::brake);
-        assert(m.to_action(sf::Keyboard::Q) == action_type::shoot);
-    }
-    {
-        const sf::Keyboard::Key key_to_go_left = sf::Keyboard::C;
-        const sf::Keyboard::Key key_to_go_right = sf::Keyboard::B;
-        const sf::Keyboard::Key key_to_accelerate = sf::Keyboard::A;
-        const sf::Keyboard::Key key_to_brake = sf::Keyboard::Z;
-        const sf::Keyboard::Key key_to_shoot = sf::Keyboard::E;
-        const key_action_map m(
-                    key_to_go_left,
-                    key_to_go_right,
-                    key_to_accelerate,
-                    key_to_brake,
-                    key_to_shoot
-                    );
-        assert(m.to_action(key_to_go_left) == action_type::turn_left);
-        assert(m.to_action(key_to_go_right) == action_type::turn_right);
-        assert(m.to_action(key_to_accelerate) == action_type::accelerate);
-        assert(m.to_action(key_to_brake) == action_type::brake);
-        assert(m.to_action(key_to_shoot) == action_type::shoot);
-    }
-    {
-        const key_action_map m = get_player_2_kam();
-        assert(m.to_action(sf::Keyboard::J) == action_type::turn_left);
-        assert(m.to_action(sf::Keyboard::L) == action_type::turn_right);
-        assert(m.to_action(sf::Keyboard::I) == action_type::accelerate);
-        assert(m.to_action(sf::Keyboard::K) == action_type::brake);
-        //assert(m.to_action(sf::Keyboard::Comma) == action_type::acc_backward);
-        assert(m.to_action(sf::Keyboard::U) == action_type::shoot);
-    }
-    {
-        const key_action_map m = get_player_1_kam();
-        assert( m.has_key(sf::Keyboard::A));
-        assert( m.has_key(sf::Keyboard::D));
-        assert( m.has_key(sf::Keyboard::W));
-        assert( m.has_key(sf::Keyboard::S));
-        assert( m.has_key(sf::Keyboard::Q));
-        assert(!m.has_key(sf::Keyboard::L));
-    }
-    #ifdef FIX_ISSUE_304
-    //Get the stun key
+  return key_action_map();
+}
+
+key_action_map get_player_2_kam()
+{
+  return key_action_map(
+        sf::Keyboard::J,
+        sf::Keyboard::L,
+        sf::Keyboard::I,
+        sf::Keyboard::K,
+        //sf::Keyboard::Comma,
+        sf::Keyboard::U
+        );
+}
+key_action_map get_player_3_kam()
+{
+  return key_action_map(
+        sf::Keyboard::Left,
+        sf::Keyboard::Right,
+        sf::Keyboard::Up,
+        sf::Keyboard::Down,
+        //sf::Keyboard::Comma,
+        sf::Keyboard::RControl
+        );
+}
+
+sf::Keyboard::Key get_random_key()
+{
+  return static_cast<sf::Keyboard::Key>(
+        static_cast<int>(sf::Keyboard::Key::A) + (std::rand() % 26)
+        );
+  // return static_cast<sf::Keyboard::Key>(std::rand() % 26);
+}
+
+key_action_map get_random_kam()
```

Comment:
Well done! This algorithm seems dauntingly powerful to me!

Reply:
Oops, I left some commented code in there 😬 
I'm going to replace lines 92-94 with line 95, does the same job but in one line and with less conversions 🌈